src/QuartoGameState.java

```diff
@@ -160,45 +165,107 @@ public void evaluate(QuartoPiece limboPiece) {
         }
         else {
             value = 0;
-            // FIXME/OPTIMIZE: this is just hacky and awful
-            for(int[] square : this.freeSquares) {
+            if(this.freeSquares.size() <= 21) {
```

Comment:
Maybe consider putting a comment here to say why we're caring about the number of free squares.